### PR TITLE
Sparse autoencoder

### PR DESCRIPTION
**Pull request by epierson9**

### CODE REVIEW

File: laplacian_variational_autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        # self.k = self.encoder_layer_sizes[-1]        
+
+        # self.decoder_layer_sizes = decoder_layer_sizes
+
+        self.initialization_function = self.glorot_init
+        self.kl_weighting = kl_weighting
+        self.non_linearity = tf.nn.sigmoid
+        self.sigma_scaling = .1
+
+    def init_network(self):
```
[the remark this answers is not in the document]
Comment (epierson9) [reply laplacian_variational_autoencoder.py]: sounds good, changed to inherit from VariationalAutoencoder

File: laplacian_variational_autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        # We assume that the prior is a Laplacian with sigma = 1, mu = 0.
+        # to compute q log p: https://www.wolframalpha.com/input/?i=integrate++1+%2F+(2+*+pi)+*+exp(-abs(x+-+7)+%2F+pi)+*+(-abs(x)+%2F+1)+from+-infinity+to+infinity
+        # to compute q log q: https://www.wolframalpha.com/input/?i=integrate++1+%2F+(2+*+pi)+*+exp(-abs(x+-+7)+%2F+pi)+*+(-abs(x+-+7)+%2F+pi)+from+-infinity+to+infinity
+        # We want to compute KL(Q, P) and we have 
+        # mu_p = 0, sigma_p = 1. Then: 
+        # KL(Q, P) = -log(sigma) - 1 - (- abs(mu) - sigma * exp(-abs(mu) / sigma))
+        # KL(Q, P) = -log(sigma) - 1 + abs(mu) + sigma * exp(-abs(mu) / sigma)
+        # which vanishes, as it should, when sigma = 1, mu = 0. 
+        
+        kl_div_loss = -tf.log(self.Z_sigma) - 1 + tf.abs(self.Z_mu) + self.Z_sigma * tf.exp(tf.abs(self.Z_mu) / self.Z_sigma)
```
[the remark this answers is not in the document]
Comment (epierson9) [reply laplacian_variational_autoencoder.py]: yeah very good catch, thanks. as i remember this error also seems consistent with the problems we were having, where the latent state was super-good at being laplacian but bad at capturing anything about the data -- we were over-penalizing deviations, i think? who knows.

File: laplacian_variational_autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        kl_div_loss = tf.reduce_mean(
+            tf.reduce_sum(
+                kl_div_loss,
+                axis=1),
+            axis=0) * self.kl_weighting
+        kl_div_loss = kl_div_loss
```
[the remark this answers is not in the document]
Comment (epierson9) [reply laplacian_variational_autoencoder.py]: just felt like making really sure that was true :) deleted.

File: laplacian_variational_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        combined_loss = binary_loss + continuous_loss + kl_div_loss
+
+        return combined_loss, binary_loss, continuous_loss, kl_div_loss  
+
+
+    def compute_elbo(self, df, continuous_variance=1):
```
[the remark this answers is not in the document]
Comment (epierson9) [reply laplacian_variational_autoencoder.py]: sounds good.

File: sparse_variational_age_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import tensorflow as tf
+from dimreducer import DimReducer
+
+from general_autoencoder import GeneralAutoencoder
+from standard_autoencoder import StandardAutoencoder
+from variational_autoencoder import VariationalAutoencoder
+from variational_age_autoencoder import VariationalAgeAutoencoder
+
+
+class SparseVariationalAgeAutoencoder(VariationalAgeAutoencoder):
```
[the remark this answers is not in the document]
Comment (epierson9) [reply sparse_variational_age_autoencoder.py]: yeah, the sparsity loss only really makes sense in that case. added an assert.
